Add support for np.rot90

braniii opened the pull request:
<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
forum https://numba.discourse.group/.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]` and
   then remove the label when you'd like it to be reviewed.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on master/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against master they should
   be resolved by merging master into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->
This PR is adds `np.rot90`, which is currently unimplemented according to #4074
The only problem is, that in numba `np.transpose` the axes argument needs to be a tuple. I am not able to create the tuple dynamically

esc commented:
@braniii thank you for submitting this. I have marked it as in-progress for now, until the transpose issues have been resolved.

stuartarchibald commented:
@braniii this should sort out the fixed tuple issue:
```patch
diff --git a/numba/np/arrayobj.py b/numba/np/arrayobj.py
index 9e81080..ea9b613 100644
--- a/numba/np/arrayobj.py
+++ b/numba/np/arrayobj.py
@@ -1569,8 +1569,10 @@ def numpy_rot90(arr, k=1):
 
     if not isinstance(k, types.Integer):
         raise errors.TypingError('The second argument "k" must be an integer')
-    if not type_can_asarray(arr):
-        raise errors.TypingError('The first argument "arr" must be array-like')
+    if not isinstance(arr, types.Array):
+        raise errors.TypingError('The first argument "arr" must be an array')
+
+    const_arr_ndim = arr.ndim
 
     def impl(arr, k=1):
         arr = np.asarray(arr)
@@ -1586,7 +1588,7 @@ def numpy_rot90(arr, k=1):
 
         axes_list = np.arange(arr.ndim)
         axes_list[:2] = [1, 0]
-        axes_list = to_fixed_tuple(axes_list, arr.ndim)
+        axes_list = to_fixed_tuple(axes_list, const_arr_ndim)
 
         if k == 1:
             return np.transpose(np.fliplr(arr), axes_list)
diff --git a/numba/tests/test_np_functions.py b/numba/tests/test_np_functions.py
index 76c7188..3f1e3e5 100644
--- a/numba/tests/test_np_functions.py
+++ b/numba/tests/test_np_functions.py
@@ -2328,7 +2328,6 @@ class TestNPFunctions(MemoryLeakMixin, TestCase):
         def a_variations():
             yield np.arange(10).reshape(5, 2)
             yield np.arange(20).reshape(5, 2, 2)
-            yield ([1, 2], [3, 4],)
 
         for a in a_variations():
             for k in range(-3, 13):
@@ -2339,7 +2338,7 @@ class TestNPFunctions(MemoryLeakMixin, TestCase):
         with self.assertRaises(TypingError) as raises:
             cfunc("abc")
 
-        self.assertIn('The first argument "arr" must be array-like',
+        self.assertIn('The first argument "arr" must be an array',
                       str(raises.exception))
 
         with self.assertRaises(TypingError) as raises:
```

braniii commented:
@stuartarchibald thank you for your help. Using your changes the test passes (at least locally). I am looking forward to a review :)

stuartarchibald commented:
> @stuartarchibald thank you for your help. Using your changes the test passes (at least locally). I am looking forward to a review :)

No problem, glad it got this working for you.

stuartarchibald reviewed (commented):
Thanks for working on implementing this feature. I've given the patch and initial review and it's a good start. There's a few Numba nuances that I've highlighted and have demoed one way to fix them in this patch https://github.com/stuartarchibald/numba/commit/85ab2e272fa9ecde0a97b5aeef3df3ac04d3986a, most of this revolves around knowing when Numba "sees" a type and what can be inferred at compile time vs. runtime. Thanks again!

braniii commented on numba/tests/test_np_functions.py:
```diff
@@ -2317,6 +2321,45 @@ def a_variations():
+            yield np.arange(20).reshape(5, 2, 2)
+
+        for a in a_variations():
+            for k in range(-3, 13):
```
I've tried to stay close to numpy's test. But you are right, it is not of need.

braniii commented on numba/np/arrayobj.py:
```diff
@@ -1562,6 +1563,39 @@ def array_T(context, builder, typ, value):
+@overload(np.rot90)
+def numpy_rot90(arr, k=1):
+    # supporting axes argument it needs to be included in np.flip
+    if not isinstance(k, types.Integer):
```
This is a valid point, I took your tests.

braniii commented on numba/np/arrayobj.py:
```diff
@@ -1562,6 +1563,36 @@ def array_T(context, builder, typ, value):
+    axes_list = (1, 0, *range(2, arr.ndim))
+
+    def impl(arr, k=1):
+        arr = np.asarray(arr)
```
Is this needed if I check for `types.Array` in line 1571? Tests passes without this line.

stuartarchibald replied:
I think that can indeed be removed.

braniii replied:
Ok, I removed it.

braniii commented:
> Thanks for working on implementing this feature. I've given the patch and initial review and it's a good start. There's a few Numba nuances that I've highlighted and have demoed one way to fix them in this patch [stuartarchibald@85ab2e2](https://github.com/stuartarchibald/numba/commit/85ab2e272fa9ecde0a97b5aeef3df3ac04d3986a), most of this revolves around knowing when Numba "sees" a type and what can be inferred at compile time vs. runtime. Thanks again!

Thank you for the detailed review and patch, it was really helpful. I changed some flake8 white space issues and simplified the tuple creation. If you prefer the explicit call of `tuple` I will revert the change.

stuartarchibald commented:
> > Thanks for working on implementing this feature. I've given the patch and initial review and it's a good start. There's a few Numba nuances that I've highlighted and have demoed one way to fix them in this patch [stuartarchibald@85ab2e2](https://github.com/stuartarchibald/numba/commit/85ab2e272fa9ecde0a97b5aeef3df3ac04d3986a), most of this revolves around knowing when Numba "sees" a type and what can be inferred at compile time vs. runtime. Thanks again!
> 
> Thank you for the detailed review and patch, it was really helpful. I changed some flake8 white space issues and simplified the tuple creation. If you prefer the explicit call of `tuple` I will revert the change.

No problem, and thanks for fixing up :)

stuartarchibald commented:
@esc @sklam this PR contains some bits of code I've authored, please could one of you take a look too, patch is pretty short. Thanks.

sklam commented on numba/np/arrayobj.py:
```diff
@@ -1562,6 +1562,35 @@ def array_T(context, builder, typ, value):
+        elif k == 3:
+            return np.fliplr(np.transpose(arr, axes_list))
+        else:
+            assert 0  # unreachable
```
i thk this should explicitly raise to prevent `python -O` causing the `assert` to be removed and falling through as a return None, which will alter the return type of the function.

braniii replied:
The `else` case is entered only when `k = k %  4` is not in `[0, 1, 2, 3]`, so when modulo operation fails. So an alternative is to change the last `elif k==3` to an `else` and  removing lines 1588-1589. In case you prefer to keep the unreachable else case, could you please help me which kind of error I should use?

sklam replied:
Putting a `raise AssertionError` is suffice. 

Details:

Since this code is compiled by numba, static analysis does not know the range of possible `k`. Type inference (a static analysis) will consider all branches in determining the return type. With `python -O`, the `assert` statement will be omitted; thus, the `else` branch becomes equivalent to a `return None`. The type-inference will then think the output type can be `None` and using an `Optional[Array]` return type unnecessarily.

braniii replied:
I've just fixed it and thx for the helpful explanation.

stuartarchibald replied:
Thanks @sklam I need to remember to stop doing this C-like idiom :)

sklam reviewed (approved):
Thanks for your contribution!

stuartarchibald commented:
@braniii Congratulations on your first contribution to Numba!
